### PR TITLE
expression, executor: rewrite builtin sha2 function

### PR DESCRIPTION
**Pull request by zs634134578**
rewrite builtin sha2 function new expression evaluation architecture

### CODE REVIEW

File: expression/builtin_encryption.go [hunk 1]
```diff
@@ -513,26 +518,18 @@ const (
 
 // eval evals a builtinSHA2Sig.
```
Comment (zz-jason) [inline expression/builtin_encryption.go]: update the comments, change `eval` to `evalString`

File: expression/builtin_encryption.go [hunk 2]
```diff
@@ -513,26 +518,18 @@ const (
...
-	}
+func (b *builtinSHA2Sig) evalString(row []types.Datum) (d string, isNull bool, err error) {
+	args := b.getArgs()
 	// Meaning of each argument:
 	// args[0]: the cleartext string to be hashed
 	// args[1]: desired bit length of result
-	bin, err := args[0].ToBytes()
-	if err != nil {
-		return d, errors.Trace(err)
+	bin, isNull, err := args[0].EvalString(row, b.ctx.GetSessionVars().StmtCtx)
```
Comment (zz-jason) [inline expression/builtin_encryption.go]: extract `b.ctx.GetSessionVars().StmtCtx` as a variable and use it in line:530

File: expression/builtin_encryption_test.go [hunk 3]
```diff
@@ -134,44 +134,81 @@ func (s *testEvaluatorSuite) TestShaEncrypt(c *C) {
...
+		f, err := newFunctionForTest(s.ctx, fcName, primitiveValsToConstants(t.args)...)
 		c.Assert(err, IsNil)
-		if tt.validCase {
-			res, err := crypt.ToString()
-			c.Assert(err, IsNil)
-			c.Assert(res, Equals, tt.crypt)
+		tp := f.GetType()
+		c.Assert(tp.Tp, Equals, mysql.TypeVarString)
+		c.Assert(tp.Charset, Equals, charset.CharsetBin)
+		c.Assert(tp.Collate, Equals, charset.CollationBin)
```
Comment (zz-jason) [inline expression/builtin_encryption_test.go]: add check for `tp.Flen` and `tp.Flag`

File: executor/executor_test.go [hunk 4]
```diff
@@ -1882,3 +1882,25 @@ func (s *testSuite) TestFuncREPEAT(c *C) {
...
+	}()
+	tk := testkit.NewTestKit(c, s.store)
+	tk.MustExec("use test")
+	// for sha2
+	tk.MustExec("drop table if exists t")
+	tk.MustExec("create table t(origin char(32), len int)")
+	tk.MustExec(`insert t values ("pingcap", 0)`)
+	tk.MustExec(`insert t values ("pingcap", 224)`)
+	tk.MustExec(`insert t values ("pingcap", 256)`)
+	tk.MustExec(`insert t values ("pingcap", 512)`)
```
Comment (zz-jason) [inline executor/executor_test.go]: add test for hash length of 384
add test for hash length out of {224, 256, 384, 512, 0}

File: expression/builtin_encryption.go [hunk 5]
```diff
@@ -494,12 +494,17 @@ type sha2FunctionClass struct {
 }
 
 func (c *sha2FunctionClass) getFunction(args []Expression, ctx context.Context) (builtinFunc, error) {
-	sig := &builtinSHA2Sig{newBaseBuiltinFunc(args, ctx)}
+	bf, err := newBaseBuiltinFuncWithTp(args, ctx, tpString, tpString, tpInt)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	types.SetBinChsClnFlag(bf.tp)
```
Comment (zz-jason) [inline expression/builtin_encryption.go]: 1. result's charset and collation should be utf8, you can just remove line:501 to fix it
2. set `bf.tp.Flen`

Comment (zs634134578) [inline expression/builtin_encryption.go]: what's a proper value for Flen? it seems depends on hash length.
Does it equal with hash length or (hash length / 4) ?

Comment (zz-jason) [inline expression/builtin_encryption.go]: `hash length / 4`

File: expression/builtin_encryption_test.go [hunk 6]
```diff
@@ -134,44 +134,81 @@ func (s *testEvaluatorSuite) TestShaEncrypt(c *C) {
...
+	}
+	fcName := ast.SHA2
+	for _, t := range cases {
+		f, err := newFunctionForTest(s.ctx, fcName, primitiveValsToConstants(t.args)...)
 		c.Assert(err, IsNil)
-		if tt.validCase {
-			res, err := crypt.ToString()
-			c.Assert(err, IsNil)
-			c.Assert(res, Equals, tt.crypt)
+		tp := f.GetType()
```
Comment (XuHuaiyu) [inline expression/builtin_encryption_test.go]: hi, a more readable test for type inferring is implemented in plan/typeinfer_test.go,
thus we can remove line 193 to line196 
and add some cases in plan/typeinfer_test.go

File: expression/builtin_encryption.go (hunk 5 above)
Comment (zs634134578) [inline expression/builtin_encryption.go]: Hi, I think I need help to get hash length in getFunction.
I tried to get it from args, but failed.
``` go
func (c *sha2FunctionClass) getFunction(args []Expression, ctx context.Context) (builtinFunc, error) {                                                
    bf, err := newBaseBuiltinFuncWithTp(args, ctx, tpString, tpString, tpInt)
    if err != nil {
        return nil, errors.Trace(err)
    }
    bfArgs := bf.getArgs()
    hashLength, isNull, err := bfArgs[1].EvalInt(nil, ctx.GetSessionVars().StmtCtx)
    if isNull || err != nil {
        return nil, errors.Trace(err)
    }
    switch int(hashLength) {
    case SHA0:
        bf.tp.Flen = SHA256 / 4
    case SHA224, SHA256, SHA384, SHA512:
        bf.tp.Flen = int(hashLength) / 4
    default:
        return nil, errors.Trace(err)
    }
    sig := &builtinSHA2Sig{baseStringBuiltinFunc{bf}}
    return sig.setSelf(sig), errors.Trace(c.verifyArgs(args))
}
```

Comment (zz-jason) [inline expression/builtin_encryption.go]: 1. You should do a type assertion, which is `const, ok := bfArgs[1].(*expression.Constant)`, before call `bfArgs[1].EvalInt(...)`.
2. If assertion succeed but the constant value is not in set {0, 224, 256 ...}, just set `Flen` to 0. **Do not** return a nil Expression in this situation
3. If assertion failed, which indicate that the parameter is not a constant value but a column reference instead, you can just set `Flen` to 512/4

File: expression/builtin_encryption.go [hunk 7]
```diff
@@ -493,7 +493,24 @@ func (c *sha2FunctionClass) getFunction(args []Expression, ctx context.Context)
 	if err != nil {
 		return nil, errors.Trace(err)
 	}
-	types.SetBinChsClnFlag(bf.tp)
+	bfArgs := bf.getArgs()
+	expectedLength := 0
+	_, ok := bfArgs[1].(*Constant)
```
Comment (XuHuaiyu) [inline expression/builtin_encryption.go]: if _, ok := xxx; ok {
}

File: expression/builtin_encryption_test.go (hunk 6 above)
Comment (XuHuaiyu) [inline expression/builtin_encryption_test.go]: This comment has not been addressed.

File: expression/builtin_encryption.go [hunk 8]
```diff
@@ -493,7 +493,24 @@ func (c *sha2FunctionClass) getFunction(args []Expression, ctx context.Context)
 	if err != nil {
 		return nil, errors.Trace(err)
 	}
-	types.SetBinChsClnFlag(bf.tp)
+	bfArgs := bf.getArgs()
+	expectedLength := 0
```
Comment (XuHuaiyu) [inline expression/builtin_encryption.go]: expectedLength := SHA512
and else in line 510 can be removed.

**Comment by XuHuaiyu** [comment]
Please resolve the conflicts.
PTAL @zz-jason

File: expression/builtin_encryption.go [hunk 9]
```diff
@@ -489,12 +489,34 @@ type sha2FunctionClass struct {
...
 func (c *sha2FunctionClass) getFunction(args []Expression, ctx context.Context) (builtinFunc, error) {
-	sig := &builtinSHA2Sig{newBaseBuiltinFunc(args, ctx)}
+	bf, err := newBaseBuiltinFuncWithTp(args, ctx, tpString, tpString, tpInt)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	bfArgs := bf.getArgs()
+	expectedLength := 0
+	_, ok := bfArgs[1].(*Constant)
+	if ok {
```
Comment (zz-jason) [inline expression/builtin_encryption.go]:
```
if _, ok := bfArgs[1].(*Constant); ok {
    ...
}
```

**Pull request closed without merging**